---
title: 'Final Project: GLANCE Study'
author: "Wendy Wang and Ning Li"
date: "Tuesday, March 10, 2015"
output:
  ioslides_presentation:
    fig_caption: yes
    fig_retina: 1
    keep_md: yes
    smaller: yes
  beamer_presentation: default
---

## Setting up some options

Let's first turn on the cache for increased performance and improved styling
```{r,cache=FALSE}
# Set some global knitr options
library("knitr")
opts_chunk$set(tidy=TRUE, size='tiny',tidy.opts=list(blank=FALSE, width.cutoff=80), cache=TRUE, messages=FALSE)
```

##Reference

**Navarro S, White E, Kantor ED, Zhang Y, Rho J, Song X, Milne GL, Lampe PD, Lampe JW. Randomized trial of glucosamine and chondroitin supplementation on inflammation and oxidative stress biomarkers and plasma proteomics profiles in healthy humans. PLoS One. 2015; 10(2):e0117534**

##Outline

- Background
- Study Design and Data Collection 
- Statistical Analysis Plan
- Data Analysis and Results
- Limitations
- Conclusions

##Background

**Glucosamine and Chondroitin (G&C)**

- One of the most popular dietary supplements in the US
- Mostly taken for osteoarthritis (OA)
- Associated with a 27-35% lower incidence of colorectal cancer, a 26-28% lower incidence of lung  cancer, 17% lower overall mortality, and a 13% lower cancer mortality.
- Safe supplements, with no known major adverse side effects.
- Regular dose
  - 1500 mg/d glucosamine hydrochloride (GHCl) + 1200 mg/d chondroitin sulfate (CS)

##Background

**Mechanisms of G&C on cancer prevention**

- Several lines of evidence from *in vitro* and preclinical studies support a possible role for G&C in reducing inflammation.
- Potentially through inhibition of nuclear factor kappa B pathway.
- No human intervention trials have been evaluated the effect of G&C on reducing inflammation or altering other pathways in healthy individuals

##Study Design

- Randomized, double-blinded, placebo-controlled crossover trial
- Ranomized on the order of treatment or placebo period
- Each intervention period lasted 28 days with a 28 day washout period in-between

##Participants

- 18 subjects: 9 males and 9 females
- Healthy, overweight(25<BMI<=32.5), non-smoking, aged 20-55 years individuals in the greater Seattle area.
- Strict exclusion criteria (e.g. medical history, medication use, large weight change, alcohol intake, supplemental use, vegetarian dietary and abnormal laboratory values)


##Data collection

- Blood samples were collected after each 28-day intervention period in the morning after a minimum of a 12-hour overnight fast.
- Blood was drawn into a tube containing EDTA for plasma. All samples were aliquoted and stored at -80 degree C.

##Proteomics Analysis

- Plasma samples were evaluated on a customized antibody array populated with ~3,000 full-length antibodies, printed in triplicates.
- Protein (200 g) from a pool of albumin and IgG-depleted plasma were labeled with Cy5. Each sample was combined with reference sample labeled with Cy3.
- Unbound proteins were removed by washing and the slides scanned for Cy3 and Cy5 fluorescence in an Axon Genepix 4000B scanner.
- The Cy5/Cy3 ratio determined the relative concentration of protein compared to reference.
- Most (>85%) antibodies on the array had coefficients of variation, for triplicates, of less than 10%. Antibodies with CV>10% were excluded for further analysis.

##Array Analysis and Normalization

- The array image was scanned using a GenePix 4000B (Axon Instruments) scanner.
- For each antibody, fold-change of the signal (red channel) was compared to the reference (green channel) after background correction, the ratio noted as M value.
- Experimental variation was normalized using within-array print-tip loess and between-array quartile normalization.
- Triplicate features were summarized using their median. M values were standardized such that the mean value and standard deviation of the placebo groups were set to zero and one, respectively.
- After all processing, data were available for analysis on a total of 2938 antibodies.

  
```{r readingdata,echo=FALSE,cache=TRUE,message=FALSE}
#load packages
library(limma)
library(qvalue)

#load our data
load("Data/ma_afterNorm_organized_18pairs.Rdata")
#ls()
#colnames(info)


#str(gene.name)
#length(unique(gene.name))
#delete any genes with more than 50% missing values
delete <- which(apply(is.na(ma$M),1,function(x) sum(x)/length(x)*100)>=50)
ma <- ma[!1:nrow(ma)%in%delete,]
#dim(ma)
#we do the same thing for gene.name
gene.name<-gene.name[!1:length(gene.name)%in%delete]
#length(rownames(ma$genes))#2938
#length(gene.name)#2938
rownames(ma)<-gene.name

#table(info$intervention)
```
##Statistical Analysis

**Overview of data**  

- We have 18 treatment and 18 placebo samples  
- We delete any genes with more than 50% missing values      
- Each has data on a 2938 antibodies  
- About 1100 of the antibodies are duplicates  

  *Would this be of concern? (Stay tuned!)*

##Statistical Analysis Plans

**Plan 1. Replicate the reported results**

- Using linear regression to adjust for batch effects
- Paired t-test comparing treatment and placebo
- GSEA analysis using Wilcoxon Test

##Statistical Analysis Plans

**Plan 2. Alternative analysis plan**

- Clean duplicated antibodies
- (Same as plan 1) Using linear regression to adjust for batch effects
- Moderated T test using R Limma package
- GSEA using GEGG and GO pathway databases and CAMERA


## Statistical Analysis Results -Plan 1
- Batch effects check(graph)
```{r batch,echo=FALSE,fig.width=8,fig.height=5,cache=TRUE}
batch.m<-ma$M

library(ggplot2)
library(reshape2)
library(gridExtra)
plotdata<-melt(batch.m)
plotdata$gender<-rep(info$gender,each=dim(batch.m)[1])
p0 <- ggplot(plotdata[plotdata$gender==0,], aes(y=Var2,x=Var1,fill=value)) + geom_tile() + scale_fill_gradient(low="white", high="darkblue")+ xlab("gene") + ylab("obs")+theme(axis.ticks = element_blank(), axis.text.y = element_blank())+ggtitle("genes expression for male")
p1 <- ggplot(plotdata[plotdata$gender==1,], aes(y=Var2,x=Var1)) + geom_tile(aes(fill=value)) + scale_fill_gradient(low="white", high="darkblue") + xlab("obs") + ylab("gene")+theme(axis.ticks = element_blank(), axis.text.y = element_blank())+ggtitle("genes expression for female")
grid.arrange( p0, p1, ncol=2)

P.Hd<-P.b<-P.p<-P.g<-NULL;
for(i in 1:dim(batch.m)[1]){
  fit1<-lm(batch.m[i,]~as.factor(info$Hybrid.day))
  fit2<-lm(batch.m[i,]~as.factor(info$box))
  fit3<-lm(batch.m[i,]~as.factor(info$position))
  fit4<-lm(batch.m[i,]~as.factor(info$gender))
  P.Hd<-c(P.Hd,anova(fit1)$"Pr(>F)"[1])
  P.b<-c(P.b,anova(fit2)$"Pr(>F)"[1])
  P.p<-c(P.p,anova(fit3)$"Pr(>F)"[1])
  P.g<-c(P.Hd,anova(fit4)$"Pr(>F)"[1])
}
Table<-matrix(0,ncol=4,nrow=2)
Table[1,1]<-mean(P.Hd<.05)#0.3158611
Table[1,2]<-mean(P.b<.05)#0.5245065
Table[1,3]<-mean(P.p<.05)#0.01565691
Table[1,4]<-mean(P.g<.05)#0.3157537
Table[2,1]<-mean(p.adjust(P.Hd, method ="bonferroni")<0.05)# 0.02791014
Table[2,2]<-mean(p.adjust(P.b, method ="bonferroni")<0.05)#0.01599728
Table[2,3]<-mean(p.adjust(P.p, method ="bonferroni")<0.05)#0
Table[2,4]<-mean(p.adjust(P.g, method ="bonferroni")<0.05)# 0.02790065
rownames(Table)<-c("prop(p-value<.05)","prop(bonferroni<.05)")
colnames(Table)<-c("Hybrid.day","box","position","gender")
```


## Statistical Analysis Results -Plan 1
- Batch effects check(multiple tests)
```{r,message=FALSE,cache=TRUE}
round(Table,digits=3)
```



## Statistical Analysis Results -Plan 1

- Using linear regression to adjust for batch effects
   *m~case+Hybrid.day+box+position+gender*     

- Paired t-test comparing treatment and placebo

- The therethold value for q value and bonferroni test  is 0.001.  
   


```{r functiondefinition,echo=FALSE,cache=TRUE}

#table(info$Hybrid.day)
sub.ma <- ma
sub.info <- info
info.temp <- sub.info
estim.casecont <- function(mylm,info.temp,m,a){
  coefs <- summary(mylm)$coef[,1]
  name.coefs <- names(coefs)
  hybed.name <- name.coefs[1:length(name.coefs) %in% grep("Hybrid.day",name.coefs)]
  hybed.name <- unlist(lapply(strsplit(hybed.name,"Hybrid.day"),function(x) x[2]))
  k <- 1
  X <- ifelse(as.character(info.temp$Hybrid.day)==hybed.name[k],1,0)
  if(length(hybed.name)>1){
    for(k in 2:length(hybed.name))  X <- cbind(X,ifelse(as.character(info.temp$Hybrid.day)==hybed.name[k],1,0))
    table(X[,ncol(X)-1],info.temp$Hybrid.day)
    table(X[,ncol(X)],info.temp$Hybrid.day)
  }
  box.name <- name.coefs[1:length(name.coefs) %in% grep("box",name.coefs)]
  box.name <- unlist(lapply(strsplit(box.name,"box"),function(x) x[2]))
  k <- 1
  X <- cbind(X,ifelse(as.character(info.temp$box)==box.name[k],1,0))
  for(k in 2:length(box.name))  X <- cbind(X,ifelse(as.character(info.temp$box)==box.name[k],1,0))
  table(X[,ncol(X)-1],info.temp$box)
  table(X[,ncol(X)],info.temp$box)
  # position #
  pos.name <- name.coefs[1:length(name.coefs) %in% grep("position",name.coefs)]
  pos.name <- unlist(lapply(strsplit(pos.name,"position"),function(x) x[2]))
  k <- 1
  X <- cbind(X,ifelse(as.character(info.temp$position)==pos.name[k],1,0))
  for(k in 2:length(pos.name))  X <- cbind(X,ifelse(as.character(info.temp$position)==pos.name[k],1,0))
  table(X[,ncol(X)-1],info.temp$position)
  table(X[,ncol(X)],info.temp$position)

  # gender #
  gen.name <- name.coefs[1:length(name.coefs) %in% grep("gender",name.coefs)]
  gen.name <- unlist(lapply(strsplit(gen.name,"gender"),function(x) x[2]))
  k <- 1
  X <- cbind(X,ifelse(as.character(info.temp$gender)==gen.name[k],1,0))
  table(X[,ncol(X)],info.temp$gender)
  
  est.m <- m - coefs[-c(1,2)]%*%t(X)
  return(est.m)
}
```

```{r,echo=FALSE,cache=TRUE}
lm.fun <- function(sub.ma,sub.info){
  sub.info$Hybrid.day <- as.factor(sub.info$Hybrid.day)
  sub.info$case <- as.factor(sub.info$case)
  sub.info$box <- as.factor(sub.info$box)
  sub.info$position <- as.factor(sub.info$position)
  sub.info$batch <- as.factor(sub.info$batch)
  sub.info$gender <- as.factor(sub.info$gender)          
  ## linear regression to M with a
  coef <- pvalue <- NA
  est.m <- matrix(rep(NA,nrow(sub.ma)*ncol(sub.ma)),nrow=nrow(sub.ma))
  i <- 1
  for(i in 1:nrow(sub.ma)){
    m <- as.numeric(sub.ma$M[i,])
    a <- as.numeric(sub.ma$A[i,])

    mylm <- lm(m ~ case +Hybrid.day+box+position+gender,data=sub.info)
    coef[i] <- mylm$coef["case1"]
    pvalue[i] <- summary(mylm)$coef["case1",4]
    est.m[i,] <- estim.casecont(mylm,sub.info,m,a)    
  }
  p.d <- data.frame(ind=1:length(pvalue),pvalue)
  q.d <- data.frame(ind=p.d$ind[!is.na(p.d$p)],q=qvalue(na.omit(p.d$p),robust=T)$qvalues)
  qvalue <- merge(p.d,q.d,by="ind",all.x=T,all.y=T)$q

  gene.name <- unlist(lapply(strsplit(as.character(sub.ma$genes$Name),"_"),function(x) x[3]))
  ab.name <- unlist(lapply(strsplit(as.character(sub.ma$genes$Name),"_"),function(x) x[4]))

  result.all <- data.frame(Name=as.character(sub.ma$genes$Name),gene.name=gene.name,ab.name=ab.name,coef=coef,pvalue=pvalue,qvalue=qvalue)
  result.all <- result.all[order(result.all$pvalue),]
#  return(result.all)
  return(list(result=result.all,est=est.m))  
}

```


```{r, fig.height=5,fig.width=8,cache=TRUE,echo=FALSE}

result <- lm.fun(ma,info)
result.all <- result$result
est.m <- result$est
rownames(est.m) <- ma$genes$Name
colnames(est.m) <- colnames(ma)

```

```{r,eval=FALSE,echo=FALSE}
par(mfrow=c(1,2),font=2)
plot(result.all$coef,result.all$pvalue,pch=19,col="blue",xlab="coef",ylab="p",main=paste(sum(info$case==1)," GC treated vs. ",sum(info$case==0)," Placebo",sep=""))
legend("topright",bty='n',legend=paste(sum(result.all$pvalue<0.05&result.all$coef>0,na.rm=T)," +\n",sum(result.all$pvalue<0.05&result.all$coef<0,na.rm=T)," - ",sep=""))
abline(h=0.05,col=6,lwd=2)
hist(result.all$pvalue,breaks=20,ylim=c(0,210),ylab=paste(sum(info$case==1)," GC treated vs ",sum(info$case==0)," Placebo",sep=""),xlab="p",main="",col="lightpink")
abline(v=0.05,col=6,lwd=2)
legend("topright",bty='n',legend=paste(sum(result.all$pvalue<0.05,na.rm=T)," signft among ",nrow(ma)," abs (", round(sum(result.all$pvalue<0.05,na.rm=T)/nrow(ma)*100,2),"%)",sep=""))
```

```{r paired t testing,echo=FALSE,cache=TRUE}

table(colnames(est.m)==info$Array.name)
info$id <- as.character(info$id)
info[1,]
table(table(info$id))
case.m <- est.m[,info$case==1]
ctrl.m <- est.m[,info$case==0]

case.info <- info[info$case==1,]
ctrl.info <- info[info$case==0,]

table(case.info$id==ctrl.info$id)

p.t <- sapply(1:nrow(case.m),function(x) t.test(case.m[x,],ctrl.m[x,],paired=T)$p.value)

effect.size <- sapply(1:nrow(case.m),function(x) (mean(case.m[x,],na.rm=T)-mean(ctrl.m[x,],na.rm=T))/sd(ctrl.m[x,],na.rm=T))
qvalue <- qvalue(p.t,robust=T)$qvalues
bonfer<-p.adjust(p.t, method = "bonferroni")
```

## Statistical Analysis Results -Plan 1

```{r Qvalueplot,echo=FALSE,cache=TRUE,fig.height=6,fig.width=6,fig.align='center'}

p.cut <- 0.01
q.cut <- 0.001

par(mfrow=c(2,2),font=6)
plot(result.all$coef,result.all$pvalue,pch=19,col="blue",xlab="coef",ylab="p",main=paste(sum(info$case==1)," GC treated vs. ",sum(info$case==0)," Placebo",sep=""))
legend("topright",bty='n',legend=paste(sum(result.all$pvalue<0.05&result.all$coef>0,na.rm=T)," +\n",sum(result.all$pvalue<0.05&result.all$coef<0,na.rm=T)," - ",sep=""))
abline(h=0.05,col=6,lwd=2)
hist(result.all$pvalue,breaks=20,ylim=c(0,210),ylab=paste(sum(info$case==1)," GC treated vs ",sum(info$case==0)," Placebo",sep=""),xlab="p",main="",col="lightpink")
abline(v=0.05,col=6,lwd=2)
legend("topright",bty='n',legend=paste(sum(result.all$pvalue<0.05,na.rm=T)," signft among ",nrow(ma)," abs (", round(sum(result.all$pvalue<0.05,na.rm=T)/nrow(ma)*100,2),"%)",sep=""))


plot(effect.size,p.t,pch=19,col="blue",xlab="Effect Size",ylab="p value paired T",main=paste(sum(info$case==1)," GC treated vs. ",sum(info$case==0)," Placebo",sep=""))
legend("topright",bty='n',legend=paste(sum(p.t<p.cut&effect.size>0,na.rm=T)," +\n",sum(p.t<p.cut&effect.size<0,na.rm=T)," - ",sep=""))
abline(h=p.cut,col=6,lwd=2)
hist(p.t,breaks=40,ylab=paste(sum(info$case==1)," GC treated vs ",sum(info$case==0)," Placebo",sep=""),xlab="p value paired T",main="",col="lightpink")
abline(v=p.cut,col=6,lwd=2)
legend("topright",bty='n',legend=paste(sum(p.t<p.cut,na.rm=T)," p<",p.cut," among ",nrow(case.m)," abs (", round(sum(p.t<p.cut,na.rm=T)/nrow(case.m)*100,2),"%)",sep=""))
```




```{r,cache=TRUE,echo=FALSE}
par(mfrow=c(2,1),font=2)
hist(qvalue,breaks=40,ylab=paste(sum(info$case==1)," GC treated vs ",sum(info$case==0)," Placebo",sep=""),xlab="p",main="",col="lightpink")
abline(v=q.cut,col=6,lwd=2)
legend("topright",bty='n',legend=paste(sum(qvalue<q.cut,na.rm=T)," q<",q.cut," among ",nrow(case.m)," abs (", round(sum(qvalue<q.cut,na.rm=T)/nrow(case.m)*100,2),"%)",sep=""))

hist(bonfer,breaks=40,ylab=paste(sum(info$case==1)," GC treated vs ",sum(info$case==0)," Placebo",sep=""),xlab="p",main="",col="lightpink")
abline(v=q.cut,col=6,lwd=2)
legend("topright",bty='n',legend=paste(sum(bonfer<q.cut,na.rm=T)," p.bonfer<",q.cut," among ",nrow(case.m)," abs (", round(mean(bonfer<q.cut,na.rm=T)*100,2),"%)",sep=""))
```

## Statistical Analysis Results -Plan 1


```{r,echo=FALSE}
out <- data.frame(Name=ma$genes$Name,effect.size,p.t,qvalue,bonfer)
out <- out[order(out$p.t),]
```

```{r,cache=TRUE}
head(out)
```

## Statistical Analysis Results -Plan 1

- GSEA analysis using Wilcoxon Test
- a Wilcoxon test to test for diserences in the distribution of p values between SNPs within the gene set under test and a control set of SNPs. 
- The Wilcoxon signed-rank test is a non-parametric statistical hypothesis test used when comparing two related samples, matched samples, or repeated measurements on a single sample to assess whether their population mean ranks differ

## Statistical Analysis Results -Plan 1

```{r,message=FALSE,echo=FALSE,cache=TRUE,fig.height=4,fig.width=5,fig.align='center'}

out$genes <- sapply(strsplit(as.character(out$Name),"_"), function(x) x[[3]])
result.save <- result <- out

#length(unique(result$genes))
result$sign.p <- ifelse(result$effect.size<0,-1,1)*(1-result$p.t)
p<-ggplot(data=result,aes(x=sign.p))+geom_histogram(colour = "darkgreen", fill = "light blue")
p

# Rearrange gene seperate by coma #
arrange <- function(data,symbols){
data.na <- data[is.na(data$genes),]
data <- data[!is.na(data$genes),]
for(i in 1:length(symbols)){
  repeat.ind <- grep(symbols[i],data$genes)
  num <- unlist(lapply(strsplit(as.character(data$genes),symbols[i]),function(x) length(x)))
  names <- unlist(strsplit(as.character(data$genes),symbols[i]))
  data <- data[unlist(sapply(1:nrow(data),function(x) rep(x,num[x]))),]
  data$genes <- names
}
data <- rbind(data.na,data)
data
}
symbols <- "/"
#dim(result)
result <- arrange(result,symbols)
#grep(symbols,result$genes)
result <- result[!is.na(result$genes) & result$genes!="",]
#dim(result)

gene.labels <- result$genes
#length(unique(result$genes))
#length(result$genes)
```

## Statistical Analysis Results -Plan 1

- GSEA analysis using GO pathways 

```{r GSEA GO1,cache=TRUE,message=FALSE,echo=FALSE}
# GO pathway on 12-16-2013 #
gs.db <- "Data/c5.all.v4.0.symbols.gmt"

db <- readLines(gs.db)
#length(db)

max.Ng <- length(db)
db.size.G <- vector(length = max.Ng, mode = "numeric") 
for (i in 1:max.Ng) {
  db.size.G[i] <- length(unlist(strsplit(db[[i]], "\t"))) - 2
}

max.size.G <- max(db.size.G)      
gs <- matrix(rep("null", max.Ng*max.size.G), nrow=max.Ng, ncol= max.size.G)
db.names <- vector(length = max.Ng, mode = "character")
db.desc <- vector(length = max.Ng, mode = "character")
num.geneset <- sapply(1:length(db),function(x) length(unique(unlist(strsplit(db[[x]], "\t"))[-c(1:2)])))
set.size <- NA
gs.count <- 1
for (i in 1:max.Ng) {
  gene.set.size <- length(unlist(strsplit(db[[i]], "\t"))) - 2
  gs.line <- noquote(unlist(strsplit(db[[i]], "\t")))
  gene.set.name <- gs.line[1] 
  gene.set.desc <- gs.line[2] 
  gene.set.tags <- vector(length = gene.set.size, mode = "character")
  for (j in 1:gene.set.size) {
    gene.set.tags[j] <- gs.line[j + 2]
  } 
  existing.set <- is.element(gene.set.tags, gene.labels)
  set.size[i] <- length(existing.set[existing.set == T])
  gs.size.threshold.min <- 25
  gs.size.threshold.max <- 500
  if (set.size[i] ==0) next
#  if ((set.size < gs.size.threshold.min) || (set.size > gs.size.threshold.max)) next
  db.size.G[gs.count] <- set.size[i]
  gs[gs.count,] <- c(gene.set.tags[existing.set], rep("null", max.size.G - db.size.G[gs.count]))
  db.names[gs.count] <- gene.set.name
  db.desc[gs.count] <- gene.set.desc
  gs.count <- gs.count + 1
} 
Ng <- gs.count - 1
gs.names <- vector(length = Ng, mode = "character")
gs.desc <- vector(length = Ng, mode = "character")
size.G <- vector(length = Ng, mode = "numeric") 
gs.names <- db.names[1:Ng]  # Gene Set Names #
gs.desc <- db.desc[1:Ng]    # Gene Set description #
size.G <- db.size.G[1:Ng]   # Number of genes in each Gene Set #
num.geneset <- num.geneset[set.size!=0]
GS <- gs[1:Ng,]
out <- NA
for(i in 1:nrow(GS)){
  x <- GS[i,]
  out <- c(out,x[x!="null"])
}
out <- out[-1]
gene.names <- unique(out) # Gene Names #
N <- length(unique(out)) # Number of unique genes #
Ns <- length(gene.labels[gene.labels%in%out]) 

print(c("Number of unique genes:", N))
print(c("Number of Gene Sets:", Ng))

X <- result$sign.p[result$genes%in%gene.names]
names(X) <- result$genes[result$genes%in%gene.names]

A <- matrix(rep(NA,Ng*length(X)),nrow=length(X))
for(i in 1:Ng){
  A[,i] <- ifelse(names(X)%in%GS[i,],1,0)
}
rownames(A) <- names(X)
colnames(A) <- gs.names

A.abnames <- as.character(result$Name[which(result$genes%in%names(X))])
#cbind(A.abnames[1:10],rownames(A)[1:10])
#table(table(A.abnames))

# ------------------ Filtering --------------- #
q <- quantile(apply(A,2,function(x) sum(x)),0.25) # at least 3 gene samples in each gene set #
select <- apply(A,2,function(x) sum(x))>q
A <- A[,select]
gs.desc <- gs.desc[select]
gs.names <- gs.names[select]
GS <- GS[select,]
num.geneset <- num.geneset[select]
# ------------------ Wilcox Test --------------- #
A <- as.data.frame(A)
#levels(factor(A[,1]))

pvalue <- sapply(1:ncol(A),function(x) wilcox.test(X~factor(A[,x]))$p.value)

# q vlaue #
qvalue <- qvalue(pvalue,robust=T)$qvalues

name <- apply(A,2,function(x) paste(unique(rownames(A)[x==1]),sep=";",collapse=";"))  # all unique gene names 
num.ab <- apply(A,2,function(x) length(unique(A.abnames[x==1])))  # num of antibody
num.gene <- apply(A,2,function(x) length(unique(rownames(A)[x==1])))  # number of genes

# summary how many genes have small pairwise t test p value #
sig.pairt <- apply(A,2,function(x) sum(result.save$p.t[which(result.save$Name%in%unique(A.abnames[x==1]))]<0.01,na.rm=T))
sig.positive <- apply(A,2,function(x) sum(result.save$effect.size[which(result.save$Name%in%unique(A.abnames[x==1]))][result.save$p.t[which(result.save$Name%in%unique(A.abnames[x==1]))]<0.01]>0))
sig.negative <- apply(A,2,function(x) sum(result.save$effect.size[which(result.save$Name%in%unique(A.abnames[x==1]))][result.save$p.t[which(result.save$Name%in%unique(A.abnames[x==1]))]<0.01]<0))
x <- A[,1]
sig.positive[1]
sig.negative[1]
sum(result.save$p.t[which(result.save$Name%in%unique(A.abnames[x==1]))]<0.01,na.rm=T)

gene.set <- data.frame(MasterGroup=seq(ncol(A)),GS.Name=colnames(A),GS.descrip=gs.desc,GS.num.gene=num.geneset,num.gene.in=num.gene,num.abprobe.in=num.ab,sig.pairt,sig.positive,sig.negative,wilc.p=pvalue,qvalue=qvalue,gene.names=name)
gene.set <- gene.set[order(gene.set$wilc.p),]

```

## Statistical Analysis Results -Plan 1

- GSEA analysis using GO pathways 

```{r}
head(gene.set)
```

## Statistical Analysis Results -Plan 1 

- Repeat GSEA analysis using KEGG pathways

```{r GESA KEGG1,cache=TRUE,echo=FALSE}

# KEGG pathway on 12-16-2013 #
gs.db <- "Data/c2.cp.kegg.v4.0.symbols.gmt"
```

```{r, echo=FALSE,message=FALSE,cache=TRUE,echo=FALSE}
db <- readLines(gs.db)
#length(db)

max.Ng <- length(db)
db.size.G <- vector(length = max.Ng, mode = "numeric") 
for (i in 1:max.Ng) {
  db.size.G[i] <- length(unlist(strsplit(db[[i]], "\t"))) - 2
}

max.size.G <- max(db.size.G)      
gs <- matrix(rep("null", max.Ng*max.size.G), nrow=max.Ng, ncol= max.size.G)
db.names <- vector(length = max.Ng, mode = "character")
db.desc <- vector(length = max.Ng, mode = "character")
num.geneset <- sapply(1:length(db),function(x) length(unique(unlist(strsplit(db[[x]], "\t"))[-c(1:2)])))
set.size <- NA
gs.count <- 1
for (i in 1:max.Ng) {
  gene.set.size <- length(unlist(strsplit(db[[i]], "\t"))) - 2
  gs.line <- noquote(unlist(strsplit(db[[i]], "\t")))
  gene.set.name <- gs.line[1] 
  gene.set.desc <- gs.line[2] 
  gene.set.tags <- vector(length = gene.set.size, mode = "character")
  for (j in 1:gene.set.size) {
    gene.set.tags[j] <- gs.line[j + 2]
  } 
  existing.set <- is.element(gene.set.tags, gene.labels)
  set.size[i] <- length(existing.set[existing.set == T])
  gs.size.threshold.min <- 25
  gs.size.threshold.max <- 500
  if (set.size[i] ==0) next
#  if ((set.size < gs.size.threshold.min) || (set.size > gs.size.threshold.max)) next
  db.size.G[gs.count] <- set.size[i]
  gs[gs.count,] <- c(gene.set.tags[existing.set], rep("null", max.size.G - db.size.G[gs.count]))
  db.names[gs.count] <- gene.set.name
  db.desc[gs.count] <- gene.set.desc
  gs.count <- gs.count + 1
} 
Ng <- gs.count - 1
gs.names <- vector(length = Ng, mode = "character")
gs.desc <- vector(length = Ng, mode = "character")
size.G <- vector(length = Ng, mode = "numeric") 
gs.names <- db.names[1:Ng]  # Gene Set Names #
gs.desc <- db.desc[1:Ng]    # Gene Set description #
size.G <- db.size.G[1:Ng]   # Number of genes in each Gene Set #
num.geneset <- num.geneset[set.size!=0]
GS <- gs[1:Ng,]
out <- NA
for(i in 1:nrow(GS)){
  x <- GS[i,]
  out <- c(out,x[x!="null"])
}
out <- out[-1]
gene.names <- unique(out) # Gene Names #
N <- length(unique(out)) # Number of unique genes #
Ns <- length(gene.labels[gene.labels%in%out]) 

print(c("Number of unique genes:", N))
print(c("Number of Gene Sets:", Ng))

X <- result$sign.p[result$genes%in%gene.names]
names(X) <- result$genes[result$genes%in%gene.names]

A <- matrix(rep(NA,Ng*length(X)),nrow=length(X))
for(i in 1:Ng){
  A[,i] <- ifelse(names(X)%in%GS[i,],1,0)
}
rownames(A) <- names(X)
colnames(A) <- gs.names

A.abnames <- as.character(result$Name[which(result$genes%in%names(X))])
#cbind(A.abnames[1:10],rownames(A)[1:10])
#table(table(A.abnames))

# ------------------ Filtering --------------- #
q <- quantile(apply(A,2,function(x) sum(x)),0.25) # at least 3 gene samples in each gene set #
select <- apply(A,2,function(x) sum(x))>q
A <- A[,select]
gs.desc <- gs.desc[select]
gs.names <- gs.names[select]
GS <- GS[select,]
num.geneset <- num.geneset[select]
# ------------------ Wilcox Test --------------- #
A <- as.data.frame(A)
#levels(factor(A[,1]))

pvalue <- sapply(1:ncol(A),function(x) wilcox.test(X~factor(A[,x]))$p.value)

# q vlaue #
qvalue <- qvalue(pvalue,robust=T)$qvalues

name <- apply(A,2,function(x) paste(unique(rownames(A)[x==1]),sep=";",collapse=";"))  # all unique gene names 
num.ab <- apply(A,2,function(x) length(unique(A.abnames[x==1])))  # num of antibody
num.gene <- apply(A,2,function(x) length(unique(rownames(A)[x==1])))  # number of genes

# summary how many genes have small pairwise t test p value #
sig.pairt <- apply(A,2,function(x) sum(result.save$p.t[which(result.save$Name%in%unique(A.abnames[x==1]))]<0.01,na.rm=T))
sig.positive <- apply(A,2,function(x) sum(result.save$effect.size[which(result.save$Name%in%unique(A.abnames[x==1]))][result.save$p.t[which(result.save$Name%in%unique(A.abnames[x==1]))]<0.01]>0))
sig.negative <- apply(A,2,function(x) sum(result.save$effect.size[which(result.save$Name%in%unique(A.abnames[x==1]))][result.save$p.t[which(result.save$Name%in%unique(A.abnames[x==1]))]<0.01]<0))
x <- A[,1]
sig.positive[1]
sig.negative[1]

sum(result.save$p.t[which(result.save$Name%in%unique(A.abnames[x==1]))]<0.01,na.rm=T)

gene.set <- data.frame(MasterGroup=seq(ncol(A)),GS.Name=colnames(A),GS.descrip=gs.desc,GS.num.gene=num.geneset,num.gene.in=num.gene,num.abprobe.in=num.ab,sig.pairt,sig.positive,sig.negative,wilc.p=pvalue,qvalue=qvalue,gene.names=name)
gene.set <- gene.set[order(gene.set$wilc.p),]
```

## Statistical Analysis Results -Plan 1 

- Repeat GSEA analysis using KEGG pathways

```{r}
head(gene.set)
```

## Statistical Analysis Results -Plan 2

- Clean duplicated antibodies

- we have 1892 genes and 36 samples
```{r delete duplicate,cache=TRUE,echo=FALSE,message=FALSE}
rm(list=ls())
library(limma)

load("E:\\UW\\winter2015\\BIOSTAT578\\project\\try\\ma_afterNorm_organized_18pairs.Rdata")
#ls()
#add is read-in from orginal dataset and output to organized data
#m is the matching indicator for ma and info$Array.name
#ls(ma)
#where ma is the dataset with expression level(M and A) and weights
#M has been normalized based on placebo distribution
#only full length of antibodies are kept in ma, ab.name and gene.name
#str(gene.name)#3586
#length(unique(gene.name))#2259
#delete genes with >50% missing data
delete <- which(apply(is.na(ma$M),1,function(x) sum(x)/length(x)*100)>=50)
#length(delete)#648
ma <- ma[!1:nrow(ma)%in%delete,]
#dim(ma)#2938
#we do the same thing for gene.name
gene.name<-gene.name[!1:length(gene.name)%in%delete]
#length(rownames(ma$genes))#2938
#length(gene.name)#2938
rownames(ma)<-gene.name
#find unique gene
unique.gene<-unique(gene.name)
#length(unique.gene)
summary.gene<-table(gene.name)
#rownames of a store all the unique gene name
#a value is count for specific gene
MA<-list(M=matrix(0,nrow=length(unique.gene),ncol=36),A=matrix(0,nrow=length(unique.gene),ncol=36))
for(i in 1:length(summary.gene)){
  name<-rownames(summary.gene)[i];
  count<-summary.gene[i];
  if(count>1){
  MA$M[i,]<-apply(ma$M[rownames(ma)==name,],2,mean,na.rm=T)
  MA$A[i,]<-apply(ma$A[rownames(ma)==name,],2,mean,na.rm=T)
  }
  else{
  MA$M[i,]<-ma$M[rownames(ma)==name,]
  MA$A[i,]<-ma$A[rownames(ma)==name,]
  }
}
#dim(MA[["M"]])
colnames(MA[["M"]])<-colnames(ma$M);
colnames(MA[["A"]])<-colnames(ma$A);
rownames(MA[["M"]])<-rownames(summary.gene);
rownames(MA[["A"]])<-rownames(summary.gene);
# we have 1892 genes and 36 samples
```

## Statistical Analysis Results -Plan 2

- (Same as plan 1) Using linear regression to adjust for batch effects
```{r, echo=FALSE,cache=TRUE}
#table(info$Hybrid.day)
info.temp <- info

info.temp$Hybrid.day <- as.factor(info.temp$Hybrid.day)
info.temp$case <- as.factor(info.temp$case)
info.temp$box <- as.factor(info.temp$box)
info.temp$position <- as.factor(info.temp$position)
info.temp$batch <- as.factor(info.temp$batch)
info.temp$gender <- as.factor(info.temp$gender)

#m <- ma$M
m<-MA$M
a<-MA$A

#remove info.temp with id=80013 and QC
info.temp <- info.temp[which(substr(info$id,1,2)!="QC"),]
#ma <- ma[,which(substr(info$id,1,2)!="QC")]
m<- m[,which(substr(info$id,1,2)!="QC")]
a<- a[,which(substr(info$id,1,2)!="QC")]
#table(info$case)

# remove 80013 since he has cold at blood draw #
select <- which(info.temp$id!="80013")
m <- m[,select]
a <- a[,select]
info.temp <- info.temp[select,]
```

```{r,echo=FALSE,cache=TRUE,echo=FALSE,message=FALSE}
#require library
library(qvalue)

estim.casecont <- function(mylm,info.temp,m,a){
  coefs <- summary(mylm)$coef[,1]
  name.coefs <- names(coefs)
  hybed.name <- name.coefs[1:length(name.coefs) %in% grep("Hybrid.day",name.coefs)]
  hybed.name <- unlist(lapply(strsplit(hybed.name,"Hybrid.day"),function(x) x[2]))
  k <- 1
  X <- ifelse(as.character(info.temp$Hybrid.day)==hybed.name[k],1,0)
  if(length(hybed.name)>1){
    for(k in 2:length(hybed.name))  X <- cbind(X,ifelse(as.character(info.temp$Hybrid.day)==hybed.name[k],1,0))
    table(X[,ncol(X)-1],info.temp$Hybrid.day)
    table(X[,ncol(X)],info.temp$Hybrid.day)
  }
  box.name <- name.coefs[1:length(name.coefs) %in% grep("box",name.coefs)]
  box.name <- unlist(lapply(strsplit(box.name,"box"),function(x) x[2]))
  k <- 1
  X <- cbind(X,ifelse(as.character(info.temp$box)==box.name[k],1,0))
  for(k in 2:length(box.name))  X <- cbind(X,ifelse(as.character(info.temp$box)==box.name[k],1,0))
  table(X[,ncol(X)-1],info.temp$box)
  table(X[,ncol(X)],info.temp$box)
  # position #
  pos.name <- name.coefs[1:length(name.coefs) %in% grep("position",name.coefs)]
  pos.name <- unlist(lapply(strsplit(pos.name,"position"),function(x) x[2]))
  k <- 1
  X <- cbind(X,ifelse(as.character(info.temp$position)==pos.name[k],1,0))
  for(k in 2:length(pos.name))  X <- cbind(X,ifelse(as.character(info.temp$position)==pos.name[k],1,0))
  table(X[,ncol(X)-1],info.temp$position)
  table(X[,ncol(X)],info.temp$position)

  # gender #
  gen.name <- name.coefs[1:length(name.coefs) %in% grep("gender",name.coefs)]
  gen.name <- unlist(lapply(strsplit(gen.name,"gender"),function(x) x[2]))
  k <- 1
  X <- cbind(X,ifelse(as.character(info.temp$gender)==gen.name[k],1,0))
  table(X[,ncol(X)],info.temp$gender)
  
  est.m <- m - coefs[-c(1,2)]%*%t(X)
  return(est.m)
}


lm.fun <- function(sub.ma,sub.info,gene.name){
  sub.info$Hybrid.day <- as.factor(sub.info$Hybrid.day)
  sub.info$case <- as.factor(sub.info$case)
  sub.info$box <- as.factor(sub.info$box)
  sub.info$position <- as.factor(sub.info$position)
  sub.info$batch <- as.factor(sub.info$batch)
  sub.info$gender <- as.factor(sub.info$gender)          
  ## linear regression to M with a
  coef <- pvalue <- NA
  #est.m <- matrix(rep(NA,nrow(sub.ma)*ncol(sub.ma)),nrow=nrow(sub.ma))
  est.m<-matrix(rep(NA,nrow(sub.ma$M)*ncol(sub.ma$M)),nrow=nrow(sub.ma$M))
  i <- 1
  for(i in 1:nrow(sub.ma$M)){
    m.temp <- as.numeric(sub.ma$M[i,])
    a.temp <- as.numeric(sub.ma$A[i,])

    mylm <- lm(m.temp ~ case +Hybrid.day+box+position+gender,data=sub.info)
    coef[i] <- mylm$coef["case1"]
    pvalue[i] <- summary(mylm)$coef["case1",4]
    est.m[i,] <- estim.casecont(mylm,sub.info,m.temp,a.temp)    
  }
  p.d <- data.frame(ind=1:length(pvalue),pvalue)
  q.d <- data.frame(ind=p.d$ind[!is.na(p.d$p)],q=qvalue(na.omit(p.d$p),robust=T)$qvalues)
  qvalue <- merge(p.d,q.d,by="ind",all.x=T,all.y=T)$q

 # gene.name <- unlist(lapply(strsplit(as.character(sub.ma$genes$Name),"_"),function(x) x[3]))
#  ab.name <- unlist(lapply(strsplit(as.character(sub.ma$genes$Name),"_"),function(x) x[4]))

  result.all <- data.frame(gene.name=gene.name,coef=coef,pvalue=pvalue,qvalue=qvalue)
  result.all <- result.all[order(result.all$pvalue),]
#  return(result.all)
  return(list(result=result.all,est=est.m))  
}

result<-lm.fun(MA,info.temp,unique.gene)
#ls(result)
est.m<-result[["est"]]
```

## Statistical Analysis Results -Plan 2

- Moderated T test using R Limma package
```{r limma,cache=TRUE}
#So we run everything in the orginal codes before paired t-test
#And we use limma instead of pairted t test to fit the model

mm <-model.matrix(~0+id+case, data=info.temp)
mm <- mm[,-11]
mm <- mm[, -(19:24)]
fit <- lmFit(est.m, mm)
```

## Statistical Analysis Results -Plan 2

- Moderated T test using R Limma package

```{r,message=FALSE}
ebay <- eBayes(fit)
top <- topTable(ebay, coef="case1", number=Inf, sort.by="p")
head(top)
mean(top$adj.P.Val<0.01)
#More than 50% genes had significant changes.
```

## Statistical Analysis Results -Plan 2

- Moderated T test using R Limma package
*Plots*
```{r, echo=FALSE,cache=TRUE,fig.align='center',fig.height=5,fig.width=8}
p.cut <- 0.01
q.cut<-0.001


case.m <- est.m[,info$case==1]
ctrl.m <- est.m[,info$case==0]

case.info <- info[info$case==1,]
ctrl.info <- info[info$case==0,]

par(mfrow=c(1,2),font=2)
plot(top$logFC,top$adj.P.Val,pch=19,col="blue",xlab="Effect Size",ylab="p value paired T",main=paste(sum(info$case==1)," GC treated vs. ",sum(info$case==0)," Placebo",sep=""))
legend("topright",bty='n',legend=paste(sum(top$adj.P.Val<p.cut&top$logFC>0,na.rm=T)," +\n",sum(top$adj.P.Val<p.cut&top$logFC<0,na.rm=T)," - ",sep=""))
abline(h=p.cut,col=6,lwd=2)

hist(top$adj.P.Val,breaks=40,ylab=paste(sum(info$case==1)," GC treated vs ",sum(info$case==0)," Placebo",sep=""),xlab="p value paired T",main="",col="lightpink")
abline(v=p.cut,col=6,lwd=2)
legend("topright",bty='n',legend=paste(sum(top$adj.P.Val<p.cut,na.rm=T)," p<",p.cut," among ",nrow(case.m)," abs (", round(sum(top$adj.P.Val<p.cut,na.rm=T)/nrow(case.m)*100,2),"%)",sep=""))
```

## Statistical Analysis Results -Plan 2

- Moderated T test using R Limma package
*Plots*

```{r,cache=TRUE,echo=FALSE}
par(mfrow=c(1,1),font=2)
hist(top$adj.P.Val,breaks=40,ylab=paste(sum(info$case==1)," GC treated vs ",sum(info$case==0)," Placebo",sep=""),xlab="p",main="",col="lightpink")
abline(v=p.cut,col=6,lwd=2)
legend("topright",bty='n',legend=paste(sum(top$adj.P.Val<p.cut,na.rm=T)," p<",p.cut," among ",nrow(case.m)," abs (", round(sum(top$adj.P.Val<p.cut,na.rm=T)/nrow(case.m)*100,2),"%)",sep=""))

```

## Statistical Analysis Results -Plan 2

- GSEA using KEGG pathway database and CAMERA
```{r,cache=TRUE,echo=FALSE}
library(GSEABase)
library(limma)
```
```{r,cache=TRUE}
kegg <- getGmt("Data/c2.cp.kegg.v4.0.symbols.gmt")
gene_ids <- geneIds(kegg)
```
```{r,cache=TRUE,echo=FALSE}
est_new <- est.m
row <- rownames(est_new)
row_new <- sapply(strsplit(as.character(row),"_"), function(x) x[[3]])
rownames(est_new) <- row_new
```
```{r,cache=TRUE}
if (exists("ids2indices")) { 
    sets_indices <- ids2indices(gene_ids, unique.gene)
}
if (exists("symbols2indices")) {
    sets_indices <- symbols2indices(gene_ids, unique.gene)    
}
cont_matrix <- makeContrasts("case1", levels=mm)
#gsea <- camera(est_new, sets_indices, design=mm, cont_matrix)
#It's not working due to the NA's
#So we change all the NA's to zero
est_narm <- est_new
est_narm[is.na(est_narm)] <- 0
gsea_kegg <- camera(est_narm, sets_indices, design=mm, cont_matrix)
```

## Statistical Analysis Results -Plan 2

- GSEA using KEGG pathway database and CAMERA

```{r,cache=TRUE}
head(gsea_kegg)
#now it worked but we don't have anything significant.
```

## Statistical Analysis Results -Plan 2

- GSEA using GO pathway database and CAMERA
```{r,cache=TRUE}
go <- getGmt("Data/c5.all.v4.0.symbols.gmt")

gene_ids_go <- geneIds(go)

if (exists("ids2indices")) { 
    sets_indices_go <- ids2indices(gene_ids_go, unique.gene)
}
if (exists("symbols2indices")) {
    sets_indices_go <- symbols2indices(gene_ids_go, unique.gene)    
}

cont_matrix <- makeContrasts("case1", levels=mm)
gsea_go <- camera(est_narm, sets_indices_go, design=mm, cont_matrix)
```

## Statistical Analysis Results -Plan 2

- GSEA using GO pathway database and CAMERA

```{r,cache=TRUE}
head(gsea_go)
#also nothing significant!
```

##Limitation

- Small sample size
- Only detecting proteints that have antibodies on the microarray


## Conclusion and Discussion

- Results for individual proteins are similar
- Results for pathway analysis are very different.

**Why?**
- We set all missing values to 0
- Different methods: paired t-test vs. LIMMA; Wilcoxon test vs. CAMERA

##Questions?


